### DISCUSSION TITLE
Windows RDP Client Error

### DISCUSSION THREAD

**Discussion by hillions** [gravitational/teleport#47705] [Q&A] [answered]
I am experimenting with connecting to a Windows Desktop using Teleport. I have successfully linked the desktop as a Resource via the Teleport Windows Desktop Service running on a Linux host. When I try to connect to the desktop via the web RDP client, I get the following error:

 `error:[the real rdpclient.Client implementation was not included in this build] event:windows.desktop.session.start`

Some brief internet searching has not revealed any solutions for this error.

`Version       16.4.3`

Any ideas?

**Comment by zmb3** [comment]
This means that RDP support is not compiled into the build you are using. 

One possible solution (assuming you meet the requirements of the license) would be to use the community edition builds that we publish on goteleport.com/download. 

If you are building from source because you don't meet the requirements of the community edition license then you'll have to rebuild teleport with the proper tooling installed.

**Comment by hillions** [comment]
I installed teleport using the install script here: https://goteleport.com/download/

```
Installing Teleport v16.4.3 through dnf
Last metadata expiration check: 22:13:42 ago on Thu Oct 17 18:23:01 2024.
Package dnf-plugins-core-4.3.0-13.amzn2023.0.4.noarch is already installed.
Dependencies resolved.
Nothing to do.
Complete!
Adding repo from: https://yum.releases.teleport.dev/amzn/2023/Teleport/x86_64/stable/v16/teleport-yum.repo
Gravitational Teleport packages                                                                                                  4.0 kB/s | 833  B     00:00
Package teleport-16.4.3-1.x86_64 is already installed.

```

**Reply by webvictim** [reply]
Could you provide the output of these commands please?

```
ls -l $(which teleport)
$(which teleport) version
sha256sum $(which teleport)
systemctl cat teleport
```

**Reply by hillions** [reply]
```
-rwxr-xr-x. 1 root root 340394336 Oct 16 11:59 /usr/local/bin/teleport
Teleport v16.4.3 git:v16.4.3-0-gd506b62 go1.22.8
3793d8a4a0b02f781feddb75760f4c8f93c65d038255ee6d02e3e920177fea44  /usr/local/bin/teleport
# /usr/lib/systemd/system/teleport.service
[Unit]
Description=Teleport Service
After=network.target

[Service]
Type=simple
Restart=always
RestartSec=5
EnvironmentFile=-/etc/default/teleport
ExecStart=/usr/local/bin/teleport start --config /etc/teleport.yaml --pid-file=/run/teleport.pid
# systemd before 239 needs an absolute path
ExecReload=/bin/sh -c "exec pkill -HUP -L -F /run/teleport.pid"
PIDFile=/run/teleport.pid
LimitNOFILE=524288

[Install]
WantedBy=multi-user.target
```

**Reply by zmb3** [reply, marked as answer]
`3793d8a4a0b02f781feddb75760f4c8f93c65d038255ee6d02e3e920177fea44` is the correct checksum for the official 16.4.3 builds, and I can confirm it _does_ include the RDP client:

```
~/Downloads/teleport ➜ sha256sum teleport
3793d8a4a0b02f781feddb75760f4c8f93c65d038255ee6d02e3e920177fea44  teleport

~/Downloads/teleport ➜ strings teleport | grep 'implementation was not'

~/Downloads/teleport took 2s ❯
```

The only things that might explain the issue you're seeing would be:

- The version of Teleport that's running on this host is not the same as whatever is located at `/usr/local/bin/teleport`. (have you tried restarting the service?)
- You have more than 1 Windows desktop agent running, and the one that's giving you this error is not the same host as the one you've grabbed the shasum from.

**Comment by hillions** [comment]
[https://github.com/gravitational/teleport/blob/master/lib/srv/desktop/rdp/rdpclient/client_nop.go](https://github.com/gravitational/teleport/blob/master/lib/srv/desktop/rdp/rdpclient/client_nop.go) seems relevant - I wonder if the desktop_access_rdp tag was used for this build.

**Reply by hillions** [reply]
And here: https://github.com/gravitational/teleport/tree/master?tab=readme-ov-file 

"If you wish to build the Rust-powered features like Desktop Access, see the Rust and Cargo versions in [build.assets/Makefile](https://github.com/gravitational/teleport/blob/master/build.assets/Makefile#L21) (search for RUST_VERSION)"

**Comment by hillions** [comment]
I believe I've resolved this issue for now. Turns out the agent which was running the Windows Desktop Service had a different build:

```
Architecture:                    armv7l
Teleport v16.4.3 git:v16.4.3-0-gd506b62 go1.22.8
f3b26c04717430cd0c3370008281d5ebbbdebed446385e0c65d0da34b5910edd  /usr/local/bin/teleport
```

Previously I had been checking the build info on the master teleport host. Changing to an x86_64 host for the Windows Desktop Service (with the corresponding build) allowed me to connect over RDP.
